toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/jsonutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/network"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/packagerepo/repocloner"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/randomization"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/retry"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/timestamp"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/pkg/profile"
+	"github.com/sirupsen/logrus"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const (
+	defaultNetOpsCount = "20"
+)
+
+type downloadResultType int
+
+const (
+	downloadResultTypeSuccess downloadResultType = iota
+	downloadResultTypeFailure
+	downloadResultTypeSkipped
+	donwloadResultTypeUnavailable
+)
+
+type downloadResult struct {
+	pkgName    string
+	resultType downloadResultType
+}
+
+var (
+	app = kingpin.New("precacher", "Pre-hydrate RPM cache for a given set of repo URLs and a RPM snapshot file.")
+
+	logFile       = exe.LogFileFlag(app)
+	logLevel      = exe.LogLevelFlag(app)
+	profFlags     = exe.SetupProfileFlags(app)
+	timestampFile = app.Flag("timestamp-file", "File that stores timestamps for this program.").String()
+
+	outDir            = exe.OutputDirFlag(app, "Directory to download packages into.")
+	snapshot          = app.Flag("snapshot", "Path to the rpm snapshot .json file.").ExistingFile()
+	outputSummaryFile = app.Flag("output-summary-file", "Path to save the summary of packages downloaded").String()
+	repoUrls          = app.Flag("repo-url", "URLs of the repos to download from.").Strings()
+	workerTar         = app.Flag("worker-tar", "Full path to worker_chroot.tar.gz").Required().ExistingFile()
+	buildDir          = app.Flag("worker-dir", "Directory to store chroot while running repo query.").String()
+
+	concurrentNetOps = app.Flag("concurrent-net-ops", "Number of concurrent network operations to perform.").Default(defaultNetOpsCount).Uint()
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	prof, err := profile.StartProfiling(profFlags)
+	if err != nil {
+		logger.Log.Warnf("Could not start profiling: %s", err)
+	}
+	defer prof.StopProfiler()
+
+	timestamp.BeginTiming("precacher", *timestampFile)
+	defer timestamp.CompleteTiming()
+
+	// Get the state of the current repo from a snapshot
+	rpmSnapshot, err := rpmSnapshotFromFile(*snapshot)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	// See what is available online from the upstream repos
+	availablePackages, err := getAllRepoData(*repoUrls, *workerTar, *buildDir)
```

Comment:
Reading "available" later in the code made me thing this is a set of packages we already know are available upstream and it was confusing to later find error logs for when we fail to actually find them in any repo. Could we re-name this variable a bit? Maybe `localPackages` or `packagesToDownload`?

Reply:
I've updated the variable to be a bit more descriptive.